documentation/CAMARA-API-access-and-user-consent.md

```diff
@@ -373,6 +377,186 @@ The Client Credentials grant type is used to obtain a 2-legged access_token that
 
 More details about the standard flow can be found in the official IETF specification [The OAuth 2.0 Authorization Framework](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4).
 
+## Offline Access - Refresh Token
+
+The defined solution uses Refesh Token as a mean to enable CAMARA API access in Offline Scenarios. Online Access means that there is a Telco-based Network Connection, either Mobile Network connection in mobile use cases or Fixed Network connection in fixed use cases.
+
+- Refresh Token may be issued as result of AuthN/AuthZ, if requested by API invoker and allowed by use case. 
+  - [RFC6749 Section 1.5](https://datatracker.ietf.org/doc/html/rfc6749#section-1.5): _A refresh token is a string representing the authorization granted to  the client by the resource owner.  The string is usually opaque to the client. The token denotes an identifier used to retrieve the authorization information. Therefore, the Refresh Token is bound to the authorization of the end user to the Application._
+
+- As indicated by [OIDC Spec](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess ), one of the uses of refresh token is allowing offline access. 
```

Comment:
I think this section is conflating things.

The `offline_access` scope is specifically defined in OpenID Connect to mean offline access to 'End-User's UserInfo Endpoint.'

This means when it receives an offline_access request, the authorization server can ask the user if they wish to grant ongoing access to the userinfo.

This same `offline_access` scope should not be extended to have the meaning of requesting equivalent ongoing access to other APIs. If there are other APIs that need to distinguish whether short or long term access is being requested those APIs should define explicit scope values with those meanings.

Reply:
@jogu (and CC @AxelNennker) That's why this usage of the `offline_access` scope is included in the proposed CAMARA OIDC profile (https://github.com/camaraproject/IdentityAndConsentManagement/pull/113), which covers the upcoming GSMA Opengateway technical stream requirement for off-net access scenarios. For example, in scenarios where the device is no longer connected to the operator's network. This information was agreed and specified in the GSMA Opengateway technical stream, and as stated in the PR description, it was agreed to move it to CAMARA.

So this is indeed an extension of `offline_access` not only to the User Info endpoint, but also for Service APIs.

https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md (release v0.1.0), containing CAMARA reached short-term agreements in this WG, also states the usage of `offline_access` for "Off-net" scenarios using refresh_token, as long as there was a connection when the first access_token was requested.

As recommended during today's WG meeting call (14/02), that we first focus on moving the necessary documentation to the CAMARA Identity & Consent WG repository as agreed. And then, if the WG decides so, we are more than happy to discuss this topic for release 0.3.0. focused on enhancements and features for the mid-term.

CC @trehman-gsma @MarkCornall

Reply:
> And then, if the WG decides so, we are more than happy to discuss this topic for release 0.3.0. focused on enhancements and features for the mid-term.

@jpengar I agree that merging the PR and then raise an issue with the needed changes could be the agreed way forward. But if there is a security issue identified and/or the `offline_access` scope can't be extended to other APIs then this MUST be corrected before v0.2.0 can be released.

Reply:
@hdamker What is exactly the security issue you mean?

Reply:
@jpengar nothing concrete as I'm not the expert here, hence the "if". I just wanted to emphasise that we can't postpone critical issues handle as enhancements and postpone to v0.3.0.

Reply:
This was the agreed solution to **support the off-net scenarios**, e.g., I get an access token when I'm using my mobile connection, then I connect to a Starbucks WiFi and can use & refresh the token. Nobody argued it constitutes a security/privacy issue (let alone critical).

It was also agreed to move the docs from the private GSMA repository, to keep a **single source of truth**, because keeping two versions of the same docs was the worst-case scenario (see #82, https://github.com/camaraproject/IdentityAndConsentManagement/issues/70#issuecomment-1778852651, https://github.com/GSMA-Open-Gateway/Open-Gateway-Documents/issues/101, https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/MeetingMinutes/2024/MOM-2024-01-17.md#redundancyoverlap-between-gsma-and-camara-doc-duplicated-discussions, etc).

Knowing that, and that  could we please merge it to get rid of the GSMA version, and then **discuss your alternatives to cover the off-net scenarios**?

Reply:
I think the document should not call "it" offline_access and then talk about refresh tokens. That are different things. (As all know)
I still think that this is wrong and should not be merged.
Something being agreed upon at GSMA does not mean that sections identified now as wrong should be merged in Camara.
I prefer to fix wrongness first and then merge.

Reply:
@AxelNennker I disagree for different reasons:
- There was discussion in the GSMA technical stream to find a solution to support off-net scenarios. This was the agreed solution. There was also an **agreement to move that doc from GSMA to CAMARA** to have a single source of truth. This is what this PR does, nothing more.
- **"Wrongness" is an opinion**. There are more solutions in the market implementing similar behaviors. For example, Okta, one of the most popular "identity as a service providers" for devs, does exactly the same
https://developer.okta.com/docs/guides/refresh-tokens/main/#get-a-refresh-token. Microsoft supports the same semantics https://learn.microsoft.com/en-us/entra/identity-platform/scopes-oidc#the-offline_access-scope. Atlassian uses offline_access exactly the same way https://developer.atlassian.com/cloud/oauth/getting-started/refresh-tokens/.  I can't judge whether this is one of those _"other contexts that are beyond the scope of this specification"_ that OIDC leaves the door open to, but it's definitely working well for thousands of devs.

I understand that you might not like it, since every solution has pros&cons (and of course we should keep improving it). But I don't think that's a good reason to block it.

Reply:
>  I can't judge whether this is one of those "other contexts that are beyond the scope of this specification" that OIDC leaves the door open to, but it's definitely working well for thousands of devs.

It's helpful to quote the full paragraph so you have the context of those 'other contexts':

> The use of Refresh Tokens is not exclusive to the offline_access use case. The Authorization Server MAY grant Refresh Tokens in other contexts that are beyond the scope of this specification.

This paragraph is making it clear that refresh tokens can be issued even when scope=offline_access is not requested. It is not suggesting that offline_access should be used in other contexts.

For example in OpenData ecosystems, refresh tokens are very commonly used, but I have never seen one that requires the offline_access scope to get a refresh token - it is common for the length of access actually needed by the client application to be indicated to the authorization server though.

As you say, some people do appear to overload offline_access to mean other things. Some people have chosen to not allow applications to indicate which resources they need ongoing access to. However given the wide range of CAMRA APIs I am not sure a decision that a refresh token should only be requested by requesting offline_access is appropriate, and I might go as far as saying that more fine grained permissioning represents current best practice.

Reply:
Is this text acceptable? I think there is no need to mention off-net or on-net. This section is taken from PR https://github.com/camaraproject/IdentityAndConsentManagement/pull/121

I guess the section should not be both in the security and interoperability profile and in CAMARA-API-access-and-user-consent. I think it should be in the "profile" and not in CAMARA-API-access-and-user-consent.

### Offline Access

Neither OIDC not OAuth2 define a way for clients to indicate whether they need a refresh_token. 
In OIDC and OAuth2 refresh token issuance is optional in Authorization Code Flow and at the discretion of the AZ.

Camara uses the scope `offline_access` in the authorization request to indicate to the AZ that the client requests a refresh token additionally to the access token for Camara API access.

---
**NOTE**

OIDC defines the scope `offline_access` and it is used to get an access token to be used at the UserInfo endpoint, and also to obtain a request token to refresh this access token.

---

The OIDC rules for `offline_access` as defined in [OIDC Offline Access](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess) apply. 

Copied to here unchanged from the OIDC standard for clarity: 

>  When offline access is requested, a prompt parameter value of consent MUST be used unless other conditions for processing the request permitting offline access to the requested resources are in place. The OP MUST always obtain consent to returning a Refresh Token that enables offline access to the requested resources. A previously saved user consent is not always sufficient to grant offline access.

> Upon receipt of a scope parameter containing the offline_access value, the Authorization Server:

>    MUST ensure that the prompt parameter contains consent unless other conditions for processing the request permitting offline access to the requested resources are in place; unless one or both of these conditions are fulfilled, then it MUST ignore the offline_access request,

>    MUST ignore the offline_access request unless the Client is using a response_type value that would result in an Authorization Code being returned,

>    MUST explicitly receive or have consent for offline access when the registered application_type is web,

>    SHOULD explicitly receive or have consent for offline access when the registered application_type is native.

The OIDC section on [OIDC Using Refresh Tokens](https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens) is not changed by this document.

The OIDC security and privacy considerations regarding offline access and refresh tokens apply e.g. on
[Token lifetime](https://openid.net/specs/openid-connect-core-1_0.html#TokenLifetime) and [Offline Access Privacy](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccessPrivacy).